Source/DafnyLanguageServer/Workspace/VerificationProgressReporter.cs

```diff
@@ -302,6 +307,12 @@ INotificationPublisher notificationPublisher
           return;
         }
 
+        if (!implementationNode.Started) {
+          // Race condition identified in https://github.com/dafny-lang/dafny/issues/3502
+          // We first need to mark the implementation as it started to verify
+          ReportVerifyImplementationRunning(implementation);
```

Comment:
Based on the Boogie code, I see how "Running" can be reporter after assertion batch results.

Looking at https://github.com/boogie-org/boogie/blob/master/Source/ExecutionEngine/ImplementationTask.cs#L163

```c#
  private async Task<VerificationResult> RunInternal(CancellationToken cancellationToken, Action<IVerificationStatus> notifyStatusChange) {

    var enqueueTask = engine.EnqueueVerifyImplementation(ProcessedProgram, new PipelineStatistics(),
      null, null, Implementation, cancellationToken, TextWriter.Null);

    // From this point on, assertion batch results can be reported, even though status has not yet been set to running (or queued even)

    var afterEnqueueStatus = enqueueTask.IsCompleted ? (IVerificationStatus)new Running() : new Queued();
    notifyStatusChange(afterEnqueueStatus);

    var verifyTask = await enqueueTask;
    if (afterEnqueueStatus is not Running) {
      notifyStatusChange(new Running());
    }

    var result = await verifyTask;

    // Assertion batch results can no longer be reported.

    CacheStatus = new Completed(result);
    notifyStatusChange(CacheStatus);
    return result;
  }
```